Bindings/Python/tests/test_states_trajectory.py

```diff
@@ -0,0 +1,243 @@
+        states = osim.StatesTrajectory.createFromStatesStorage(
+                model, sto, False, True)
+        states = osim.StatesTrajectory.createFromStatesStorage(
+                model, sto, True, False)
+        states = osim.StatesTrajectory.createFromStatesStorage(
+                model, sto, True, True)
+
+    def test_populate_trajectory(self):
+        model = osim.Model(os.path.join(test_dir,
+            "gait10dof18musc_subject01.osim"))
+        state = model.initSystem()
+        states = osim.StatesTrajectory()
+        states.append(state)
+        state.setTime(1.0)
+        states.append(state)
+
+        self.assertEqual(states.getSize(), 2)
+        self.assertEqual(states[1].getTime(), 1.0)
+
+    def test_out_of_range(self):
+        model = osim.Model(os.path.join(test_dir,
+            "gait10dof18musc_subject01.osim"))
+        state = model.initSystem()
+        states = osim.StatesTrajectory()
+        states.append(state)
+        state.setTime(1.0)
+        states.append(state)
+
+        # TODO this exception message could be better...
+        self.assertRaises(RuntimeError, lambda: states[2].getTime())
+
+    def test_integrity_checks(self):
+        model = osim.Model(os.path.join(test_dir,
+            "gait10dof18musc_subject01.osim"))
+        state = model.initSystem()
+        states = osim.StatesTrajectory()
+        states.append(state)
+        state.setTime(1.0)
+        states.append(state)
+        self.assertTrue(states.isNondecreasingInTime())
+        self.assertTrue(states.isConsistent())
+        self.assertTrue(states.hasIntegrity())
+
+        # Cannot append a state with an earlier time than the last one.
+        state.setTime(0.5)
+        self.assertRaises(RuntimeError, states.append, state)
+
+        # However, since python doesn't have constness, we can edit the time of
+        # a state in the trajectory.
+        state.setTime(1.5)
+        states.append(state)
+        self.assertTrue(states.isNondecreasingInTime())
+        states.back().setTime(0.25)
+        self.assertFalse(states.isNondecreasingInTime())
+        self.assertTrue(states.isConsistent())
+        self.assertFalse(states.hasIntegrity())
+
+        # TODO check violating isConsistent() (might need a different model).
+    
+    def test_access_by_time(self):
+
+        model = osim.Model(os.path.join(test_dir,
+            "gait10dof18musc_subject01.osim"))
+        state = model.initSystem()
+        states = osim.StatesTrajectory()
+        state.setTime(0.5)
+        states.append(state)
+        state.setTime(0.6)
+        states.append(state)
+        state.setTime(0.7)
+        states.append(state)
+        state.setTime(0.8)
+        states.append(state)
+        state.setTime(0.9)
+        states.append(state)
+        state.setTime(0.9)
+        states.append(state)
+        state.setTime(0.9)
+        states.append(state)
+        state.setTime(1.2)
+        states.append(state)
+
```

Comment:
> Isn't the use case that actual states are truncated and have a loss of precision (e.g. 4.99999999999) but very close to user intended start/end times that are crisp/precise (like 5)?

Yes, I think so.

> I am not convinced the tolerance works as expected in that case where you would allow some slop above or below the truncated value as a match to the user crisp value.

I think the tolerance works in this case (that, or I don't understand this scenario). I'll respond to this in more detail in the main thread.